### ISSUE TITLE
Python script include a module library from google-api-python-client [1]

### ISSUE THREAD

**Issue by mbettan** [CloudSlang/cloud-slang#449]
Hello,

In order to be able to authenticate to Google APIs, I would like create a Python Script action using google-api-python-client provided by google for Python library (https://developers.google.com/api-client-library/python/start/installation)

How could I include google-api-python-client library inside CloudSlang please ? Any ideas ?

More details below 

After having installed google-api-python-client correctly in laptop, I can execute a python standalone script to do the authentification 
```python
#!/usr/bin/env python
import argparse
import json
import os

from apiclient import discovery
from oauth2client.client import GoogleCredentials
os.environ['GOOGLE_APPLICATION_CREDENTIALS'] = 'C:\\Users\\bettanm\\Documents\\GitHub\\cloud-slang-content\\content\\io\\cloudslang\\cloud_provider\\gke\\cloudslang-026ac0ebb6e0.json'
credentials = GoogleCredentials.get_application_default()
service = discovery.build('container', 'v1', credentials=credentials)
```
Unfortunately, it's not working with a YAML CloudSlang file 
```yaml
namespace: io.cloudslang.cloud_provider.gke

operation:
  name: google_auth
  inputs:
    - jSonPathauthentification:
        required: false
  action:
    python_script: |
        try:
            import argparse
            import json
            import os

            from apiclient import discovery
            from oauth2client.client import GoogleCredentials
            os.environ['GOOGLE_APPLICATION_CREDENTIALS'] = 'C:\\Users\\bettanm\\Documents\\GitHub\\cloud-slang-content\\content\\io\\cloudslang\\cloud_provider\\gke\\cloudslang-026ac0ebb6e0.json'
            credentials = GoogleCredentials.get_application_default()
            service = discovery.build('container', 'v1', credentials=credentials)
        except Exception as ex:
            return_code = '-1'
            return_result = ex
  outputs:
    - return_code
    - return_result
    - error_message: return_result if return_code == '-1' else ''

  results:
    - SUCCESS: return_code == '0'
    - FAILURE
```
Thanks very much
Michaël B.

**Comment by oritstone** [comment]
@mbettan what is the problem you encounter? 
please attach the execution.log of the CLI

**Comment by meirwah** [comment]
Hi @mbettan 
did you look at this : http://cloudslang-docs.readthedocs.org/en/v0.8/tutorial/14_lesson/   ?

**Comment by mbettan** [comment]
Thanks @meirwah, I have modified requirements.txt to include following 

```
setuptools
jenkinsapi == 0.2.22
google-api-python-client == 1.4.2
uritemplate == 0.6
six == 1.9.0
oauth2client == 1.5.1
httplib2 == 0.9.1
```

Unfortunately, it's still not working, I have following error when executing through CloudSlang
```
Command failed java.lang.RuntimeException: Slang Error : Error executing python
script: Traceback (most recent call last):
  File "<string>", line 7, in <module>
  File "C:\cslang\bin\..\python-lib\apiclient\__init__.py", line 19, in <module>

    from googleapiclient import discovery
  File "C:\cslang\bin\..\python-lib\googleapiclient\discovery.py", line 68, in <
module>
    from oauth2client.client import GoogleCredentials
  File "C:\cslang\bin\..\python-lib\oauth2client\client.py", line 1674, in <modu
le>
    _cached_http = httplib2.Http(MemoryCache())
AttributeError: 'module' object has no attribute 'Http'
```

Note that it is following working when I execute it through Python 3 on my machine without CloudSlang
```
import argparse
import json
import os

from apiclient import discovery
from oauth2client.client import GoogleCredentials
os.environ['GOOGLE_APPLICATION_CREDENTIALS'] = 'C:\\Users\\bettanm\\Documents\\GitHub\\cloud-slang-content\\content\\io\\cloudslang\\cloud_provider\\gke\\cloudslang-026ac0ebb6e0.json'
credentials = GoogleCredentials.get_application_default()
service = discovery.build('container', 'v1', credentials=credentials)
```

@oritstone - Below are the full logs and cloudslang file
http://we.tl/5GjKRYxKn5

Any ideas how to solve this issue in CloudSlang to have this module dependency ? Does CloudSlang use Python 2 and not Python 3?

Thanks for your help

**Comment by meirwah** [comment]
@mbettan we are supporting Python 2.7 (no 3), and we have a limitation that we do not support Python libs that need CPython (It's a Jython limitation actually)...
Need to check if your libs are using CPython...

**Comment by mbettan** [comment]
@meirwah @oritstone  - I don't think that Google API Client require Cython
https://github.com/google/google-api-python-client
https://github.com/google/oauth2client

Any way to double check it through command line?

**Comment by mbettan** [comment]
Google github project proposed following workaround - https://github.com/google/google-api-python-client/issues/136

If I patch httplib2/sock.py as follows
```
# import socket
from socket import socket
```

This is fully working now, but how to manage this automatic fix with CloudSlang ? pip will automatically install the original module for each new deployments.

Any ideas ?

**Comment by meirwah** [comment]
maybe fork google-api-python-client, and create your version of it (with the fix)?
then use that in CloudSlang?

**Comment by mbettan** [comment]
Google Engineers raised an issue and pull request to the dependency httplib2 (https://github.com/jcgregorio/httplib2/pull/300) to fix this Jython compatibility. I hope it will be solved soon.

I will add into "cloud-slang-content\python-lib\requirements.txt" following dependency
```
google-api-python-client
```
google-api-python-client is dependent from httplib2, so it will automatically download the original library.

I suggest to commit this specific folder in the cloudslang repository in "cloud-slang-content\python-lib\httplib2"  - pip install should not override it right ?

Do you think it will work ? Otherwise, can I fork httplib2 and force cloudslang to download my specific hotfix version?

**Comment by meirwah** [comment]
I think it makes more sense to fork httplib2 + google-api-python-client, and make CS use your version of the fork (with the hotfix)... then, when Google will fix this bug, we can easily switch from your fork version to the original master version...

**Comment by mbettan** [comment]
I don't see how CloudSlang can use my  httplib2 + google-api-python-client fork, as "requirements.txt" desribes package to install through pip install command line.

pip install would only use official content right? What am I missing here?

Thanks, but I agree on your proposal

**Comment by orius123** [comment]
Does this help? http://stackoverflow.com/questions/20101834/pip-install-from-github-repo-branch

**Comment by mbettan** [comment]
Google engineer fixed the issue with the new httplib 0.9.2 release
Problem solved with Jython
